nack/responder_interceptor.go

```diff
@@ -0,0 +1,127 @@
+	interceptor.NoOp
+	size    uint16
+	streams *sync.Map
+	log     logging.LeveledLogger
+}
+
+type localStream struct {
+	sendBuffer *sendBuffer
+	rtpWriter  interceptor.RTPWriter
+}
+
+// NewResponderInterceptor returns a new GeneratorInterceptor interceptor
+func NewResponderInterceptor(opts ...ResponderOption) (*ResponderInterceptor, error) {
+	r := &ResponderInterceptor{
+		NoOp:    interceptor.NoOp{},
+		size:    8192,
+		streams: &sync.Map{},
+		log:     logging.NewDefaultLoggerFactory().NewLogger("nack_responder"),
+	}
+
+	for _, opt := range opts {
+		opt(r)
+	}
+
+	_, err := newSendBuffer(r.size)
+	if err != nil {
+		return nil, err
+	}
+
+	return r, nil
+}
+
+// BindRTCPReader lets you modify any incoming RTCP packets. It is called once per sender/receiver, however this might
+// change in the future. The returned method will be called once per packet batch.
+func (n *ResponderInterceptor) BindRTCPReader(reader interceptor.RTCPReader) interceptor.RTCPReader {
+	return interceptor.RTCPReaderFunc(func() ([]rtcp.Packet, interceptor.Attributes, error) {
+		pkts, attr, err := reader.Read()
+		if err != nil {
+			return nil, nil, err
+		}
+
+		for _, rtcpPacket := range pkts {
+			nack, ok := rtcpPacket.(*rtcp.TransportLayerNack)
+			if !ok {
+				continue
+			}
+
+			go n.resendPackets(nack)
+		}
+
+		return pkts, attr, err
+	})
+}
+
+// BindLocalStream lets you modify any outgoing RTP packets. It is called once for per LocalStream. The returned method
+// will be called once per rtp packet.
+func (n *ResponderInterceptor) BindLocalStream(info *interceptor.StreamInfo, writer interceptor.RTPWriter) interceptor.RTPWriter {
+	hasNack := false
+	for _, fb := range info.RTCPFeedback {
+		if fb.Type == "nack" && fb.Parameter == "" {
+			hasNack = true
+		}
+	}
+
+	if !hasNack {
+		return writer
+	}
+
+	// error is already checked in NewGeneratorInterceptor
+	sendBuffer, _ := newSendBuffer(n.size)
+	n.streams.Store(info.SSRC, &localStream{sendBuffer: sendBuffer, rtpWriter: writer})
+
+	return interceptor.RTPWriterFunc(func(p *rtp.Packet, attributes interceptor.Attributes) (int, error) {
+		sendBuffer.add(p)
+
+		return writer.Write(p, attributes)
+	})
+}
+
+// UnbindLocalStream is called when the Stream is removed. It can be used to clean up any data related to that track.
+func (n *ResponderInterceptor) UnbindLocalStream(info *interceptor.StreamInfo) {
+	n.streams.Delete(info.SSRC)
+}
+
+func (n *ResponderInterceptor) resendPackets(nack *rtcp.TransportLayerNack) {
+	v, ok := n.streams.Load(nack.MediaSSRC)
+	if !ok {
+		return
+	}
+
+	stream := v.(*localStream)
+	seqNums := nackParsToSequenceNumbers(nack.Nacks)
+
+	for _, seq := range seqNums {
+		p := stream.sendBuffer.get(seq)
+		if p == nil {
+			continue
+		}
+
+		_, err := stream.rtpWriter.Write(p, interceptor.Attributes{})
+		if err != nil {
+			n.log.Warnf("failed resending nacked packet: %+v", err)
+		}
+	}
+}
+
+func nackParsToSequenceNumbers(pairs []rtcp.NackPair) []uint16 {
```

Comment:
```suggestion
func nackPairsToSequenceNumbers(pairs []rtcp.NackPair) []uint16 {
```

Reply:
Removed function entirely, used similar thing in `pion/rtcp`